File: src/main/java/io/github/foodjournal/messaging/InboundMessage.java
package io.github.foodjournal.messaging;

import java.util.List;

/** The only message shape the journal processing pipeline accepts. */
public record InboundMessage(String provider, String eventId, String userId, String conversationId,
    String displayName, String languageCode, String text, String caption, List<Attachment> attachments) {
  public InboundMessage { attachments = attachments == null ? List.of() : List.copyOf(attachments); }
}

File: python-app/app/messaging/inbound_message.py
from dataclasses import dataclass, field
from enum import Enum


class AttachmentKind(str, Enum):
    VOICE = "VOICE"
    PHOTO = "PHOTO"
    DOCUMENT = "DOCUMENT"


@dataclass(frozen=True)
class Attachment:
    kind: AttachmentKind
    handle: str
    mime_type: str | None
    file_name: str | None = None


@dataclass(frozen=True)
class InboundMessage:
    """The only message shape the journal processing pipeline accepts, regardless
    of which provider (telegram/mattermost/terminal) it came from."""

    provider: str
    event_id: str
    user_id: str
    conversation_id: str
    display_name: str | None
    language_code: str | None
    text: str | None
    caption: str | None
    attachments: list[Attachment] = field(default_factory=list)
